snuba/subscriptions/consumer.py

```diff
@@ -0,0 +1,256 @@
+        self.__consumer = KafkaConsumer(
+            {**configuration, "enable.auto.offset.store": "false"}
+        )
+
+        # XXX: There's no way to deal with this right now and *not* store
+        # offsets, so this option is effectively required.
+        self.__enable_auto_offset_store = get_boolean_configuration_value(
+            configuration, "enable.auto.offset.store", True
+        )
+
+    def subscribe(
+        self,
+        topics: Sequence[str],
+        on_assign: Optional[Callable[[Sequence[TopicPartition]], Any]] = None,
+        on_revoke: Optional[Callable[[Sequence[TopicPartition]], Any]] = None,
+    ) -> None:
+        def on_assign_callback(
+            consumer: Any, assignment: Sequence[TopicPartition]
+        ) -> None:
+            for tp in assignment:
+                key = (tp.topic, tp.partition)
+                if key not in self.__topic_partition_states:
+                    state = PartitionState()
+                    self.__topic_partition_states[key] = state
+                    logger.debug("Initialized partition state for %r: %r", key, state)
+
+            if on_assign is not None:
+                on_assign(assignment)
+
+        def on_revoke_callback(
+            consumer: Any, assignment: Sequence[TopicPartition]
+        ) -> None:
+            # XXX: Check to see if this causes any weird shit during rebalancing.
+            for tp in assignment:
+                key = (tp.topic, tp.partition)
+                state = self.__topic_partition_states.pop(key)
+                logger.debug("Discarded partition state for %r: %r", key, state)
+
+            if on_revoke is not None:
+                on_revoke(assignment)
+
+        self.__consumer.subscribe(
+            topics, on_assign=on_assign_callback, on_revoke=on_revoke_callback
+        )
+
+    def poll(self, timeout: Optional[float] = None) -> Optional[TaskSet]:
+        message = self.__consumer.poll(*[timeout] if timeout is not None else [])
+        if message is None:
+            return None
+
+        error = message.error()
+        if error is not None:
+            raise error
+
+        timestamp_type, timestamp = message.timestamp()
+        assert timestamp_type == TIMESTAMP_LOG_APPEND_TIME
+
+        state = self.__topic_partition_states[
+            (message.topic(), message.partition())
+        ].send(Position(message.offset(), timestamp / 1000.0))
+
+        if isinstance(state, Position):
+            tasks = None
+        elif isinstance(state, Interval):
+            # TODO: Fetch tasks that are scheduled between the interval
+            # endpoints for this topic and partition.
+            tasks = TaskSet(message.topic(), message.partition(), state, [])
+        else:
+            raise ValueError("unexpected state")
+
+        if self.__enable_auto_offset_store:
+            self.__consumer.store_offsets(
+                offsets=[
```

Comment:
This should work as expected, and I think there is probably some confusion due to the difference in the way that the Kafka driver handles offset storage and committed message offsets (which I perpetuated in my long comment that should probably be rewritten to be clearer.)

The important thing to note is that committed offset for a partition is the first offset that the a consumer will receive from that partition after restarting or rebalancing. This is different than committing a _message_, which is to commit that message's offset + 1 so that that same message isn't fetched again by a consumer in that partition.

This example using [this test code](https://gist.github.com/tkaemming/6ed94da5ae9dd4b352ff626006d24584 ) should demonstrate this behavior more clearly (I deleted some logging to make it a bit easier to understand):

```python
>>> produce('example-topic')
>>> consume('example-topic')
no message
example-topic 0 0 b'0'
example-topic 0 1 b'1'
example-topic 0 2 b'2'
example-topic 0 3 b'3'
example-topic 0 4 b'4'
no message
commit [TopicPartition{topic=example-topic,partition=0,offset=4,error=None}]
no message
^C
```

At this point, messages 0 through 3 (inclusive) have been committed, message 4 has not been committed, but the committed offset is the same as the offset of message 4. Restarting the consumer will then resume with message 4:

```python
>>> consume('example-topic', 5)
no message
example-topic 0 4 b'4'
example-topic 0 5 b'5'
example-topic 0 6 b'6'
example-topic 0 7 b'7'
example-topic 0 8 b'8'
no message
^C
```

So, using the Interval(MA, MB) example again on cold start:

- First call to `poll`: the consumer receives MA, the consumer stores the offset of MA, the consumer returns None.
  - If at this point we either commit offsets or crash, we'll restart exactly where we left off: the first message consumed will again be MA.
- First call to `poll`: the consumer receives MB, the consumer stores the offset of MB, the consumer returns the task set for Interval(MA, MB).
  - If we do not have auto commit (`enable.auto.commit`) enabled and crash here, we'll restart with MA.
  - If we do not have auto commit enabled and we explicitly commit the stored offset (e.g. after the tasks are completed) and crash here, we'll restart with MB.
  - If we _do_ have auto commit enabled and crash here, one of those two things will happen indeterminately.

Reply:
ok, understood. Makes sense.
Though my followup is whether the behavior is the same if someone commits after receiving a task list by calling the commit method. In this case you are specifying the offset, thus you are right, you will restart from that offset you saved. But in the commit case you are not specifying an offset. The documentation says

```
message and offsets are mutually exclusive, if neither is set the current partition assignment’s offsets are used instead
```

So, in this case what is "the current partition assignment" ? Is it the offset of the last consumed message + 1 or is it the last offset you stored with store_offset ?
If calling commit after getting the taskSet between MA and MB commits MB+1, that would be the upper bound. But if it is the last offset you stored then it works.

Reply:
> So, in this case what is "the current partition assignment" ? Is it the offset of the last consumed message + 1 or is it the last offset you stored with store_offset ?

My understanding is that it should be the latter (the last offset you stored with `store_offset`), but I'll double check.

Reply:
Ok, I just reran the same script with autocommit disabled and an explicit synchronous commit in the `except KeyboardInterrupt` clause:

```diff
diff --git a/test.py b/test.py
index 7fcdd1b..cdf3199 100644
--- a/test.py
+++ b/test.py
@@ -22,7 +22,7 @@ def consume(topic, n=5):
             {
                 "group.id": "test",
                 "auto.offset.reset": "earliest",
-                "enable.auto.commit": "true",
+                "enable.auto.commit": "false",
                 "enable.auto.offset.store": "false",
                 "default.topic.config": {"auto.offset.reset": "earliest"},
                 "on_commit": on_commit,
@@ -56,4 +56,5 @@ def consume(topic, n=5):
             if i >= n:
                 consumer.pause([TopicPartition(message.topic(), message.partition())])
     except KeyboardInterrupt:
+        consumer.commit(asynchronous=False)
         consumer.close()
```

It worked as expected (except apparently that doesn't trigger the `commit_cb`):

```python
>>> consume('example')
no message
example 0 10 b'0'
example 0 11 b'1'
example 0 12 b'2'
example 0 13 b'3'
example 0 14 b'4'
no message
^C
```

```python
>>> consume('example')
no message
example 0 14 b'4'
example 0 15 b'5'
example 0 16 b'6'
example 0 17 b'7'
example 0 18 b'8'
^C
```

Reply:
Also, per the documentation: 

> The stored offsets will be committed according to ‘auto.commit.interval.ms’ or manual offset-less commit().